File: Topics/Design/HitCounter.java
package Topics.Design;

import java.util.ArrayDeque;
import java.util.Deque;

public class HitCounter {
    Deque<int[]> q;
    int size;

    /** Initialize your data structure here. */
    public HitCounter() {
        q = new ArrayDeque<>();
    }

    /** Record a hit.
     @param timestamp - The current timestamp (in seconds granularity). */
    public void hit(int timestamp) {
        if (q.isEmpty() || q.peekLast()[0] != timestamp) {
            q.offer(new int[]{timestamp, 1});
        } else {
            q.peekLast()[1]++;
        }
        size++;
    }

    /** Return the number of hits in the past 5 minutes.
     @param timestamp - The current timestamp (in seconds granularity). */
    public int getHits(int timestamp) {
        while (!q.isEmpty() && q.peekFirst()[0]+300 <= timestamp) {
            size -= q.pollFirst()[1];
        }
        return size;
    }
}

File: Topics/Design/HitCounter.cpp
// You need to design a hit counter system that tracks the number of hits received within the past 5 minutes (300 seconds).

// Recording hits: When a hit occurs at a specific timestamp (in seconds), the system should record it. 
// Multiple hits can happen at the same timestamp.

// Querying hit count: Given a timestamp, the system should return the total number of hits that occurred in the past 300 seconds from that timestamp. 
// Specifically, it counts all hits in the time range [timestamp - 299, timestamp].

class HitCounter {
public:
    HitCounter() : totalHits(0) {}

    void hit(int timestamp) {
        // lock_guard<mutex> lock(mutex_);
        
        if (hits.empty() || hits.back().first != timestamp) {
            hits.push_back({timestamp, 1});
        } else {
            ++hits.back().second;
        }

        ++totalHits;
    }

    int getHits(int timestamp) {
        // lock_guard<mutex> lock(mutex_);
        
        while (!hits.empty() &&
               hits.front().first + 300 <= timestamp) {
            totalHits -= hits.front().second;
            hits.pop_front();
        }

        return totalHits;
    }

private:
    deque<pair<int, int>> hits; // {timestamp, count}
    int totalHits;
    // mutex mutex_;
};
